### PR TITLE
fix: close unclosed tags in html parser and trim indentation of md content lines

### PR DESCRIPTION
**Pull request by razortanmayshah**
# Changes Made : 
1. Extend md utitlity to close unclosed tags.
    **Reason** : In Citadel's markdown to html parser, some tags were left unclosed which generates incorrect html syntax. This generated html is used to build Algolia records.
Until now, we were not doing HAST parsing while building Algolia records, hence, this error went unnoticed. Now, we are parsing HAST for generating granular search records. Hence need this fix.

2. Trim indentation of md content before rendering to html
     **Reason**: if content is not trimmed, then this results in incorrect interpretation of indentation resulting in errors like adding codeblocks whenver indentation > 3 etc. One more benefit of this is that record size is decreased by a great extent. No of heaving records decreased from 12 to 4.

### CODE REVIEW

File: scripts/md.js [hunk 1]
```diff
@@ -112,12 +167,62 @@ function getMd(config) {
...
     };
   };
 };
+
+/**
+ * Generate unique slug for each heading on the page
+ * @param {Object} listOfSlugs map of all slugs
+ * @returns {Function} function to generate unique slug
+ */
+function createSlug(listOfSlugs) {
```
Comment (PranjalMaithani) [inline scripts/md.js:213]: suggestion: listOfSlugs can be created inside the `createSlug` function directly if it's not being used outside.
Can also rename this to `slugCountMap` if the mapping is like this 

```
{
    slugA: 1,
    slugB: 2
}
```

File: scripts/md.js [hunk 2]
```diff
@@ -112,12 +167,62 @@ function getMd(config) {
...
+    openTags.reduceRight((_, openTag) => {
+      content += `</${openTag}>`;
+      return _;
+    }, {});
+
+    
+
+    /**
+     * sanitizeHtml function is used here to clean up the html generated with unclosed / wrongly closed tags
+     */
```
Comment (PranjalMaithani) [inline scripts/md.js:187]: Possible to add an example of how the before and after might look like?

File: build-search.js [hunk 3]
```diff
@@ -52,14 +52,19 @@ const build = async (config) => {
 
   docsWithContent.forEach((doc) => {
     let content = atRules(doc.body, config, doc.key);
-    const parsedContent = markdown(content, config);
+    // trim the content to remove extra empty lines and spaces
+    // if content is not trimmed, then this results in incorrect interpretation of indentation resulting
+    // resulting in errors like adding codeblocks whenver indentation > 3 etc.
```
Comment (PranjalMaithani) [inline build-search.js:57]: Question: For codeblocks to get added is it happening with indentation > 3? Assuming there should be backticks also (```)

File: plugins.js [hunk 4]
```diff
@@ -95,8 +95,8 @@ function initializePlugin(plugin, config) {
   plugin.init(config);
 }
 
-function applyPlugin({ plugin, content }) {
-  const result = plugin.process(content);
+function applyPlugin({ plugin, content, config }) {
+  const result = plugin.process(content, config);
```
Comment (PranjalMaithani) [inline plugins.js:99]: Question: Is passing config required?

File: build-search.js [hunk 5]
```diff
@@ -52,14 +52,19 @@ const build = async (config) => {
...
+    const trimmedContent = content.split('\n').map(line => line.trim()).join('\n')
+    const parsedContent = markdown(trimmedContent, config);
     const parsedContentWithFrontmatter = {
       ...parsedContent,
       ...doc.frontMatter,
       body: content,
       href: doc.href,
+      content: parsedContent.content.split('\n').join(''),
```
Comment (PranjalMaithani) [inline build-search.js:65]: Question: Why is this required? Is this another thing which was missed earlier?

Comment (razortanmayshah) [reply build-search.js:65]: Will add a comment on this

File: build-search.js (hunk 3 above)
Comment (razortanmayshah) [reply build-search.js:57]: We are ourselves treading indentation > 3 as code blocks. Check this code : https://github.com/razorpay/citadel/blob/DX-574-improve-search-experience-for-faq-and-tabs/scripts/html_block.js#L40

File: build-search.js (hunk 5 above)
Comment (razortanmayshah) [reply build-search.js:65]: This is done to remove the extra spacing and trim down the html dom string

File: plugins.js (hunk 4 above)
Comment (razortanmayshah) [reply plugins.js:99]: yes.

File: scripts/md.js (hunk 2 above)
Comment (razortanmayshah) [reply scripts/md.js:187]: Have added the reference to the github package

File: scripts/at-rules.js [hunk 6]
```diff
@@ -82,7 +82,7 @@ module.exports = function (body, config, docPath) {
       function (_, indent, openTag, content) {
         attrs = getAttrs(openTag, docPath, config);
 
-        if (config.org === attrs.org) return `${indent}${content}`;
+        if (config.org === attrs.org && config.countryCode === attrs.country ) return `${indent}${content}`;
```
[the remark this answers is not in the document]
Comment (razortanmayshah) [reply scripts/at-rules.js:85]: support for this will be added later. For now, this attribute is only used for razorpay-IN org and axis org.

File: build-search.js (hunk 5 above)
Comment (PranjalMaithani) [reply build-search.js:65]: Please add this as a comment in code

File: scripts/at-rules.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (PranjalMaithani) [reply scripts/at-rules.js:85]: org-country attribute has been used for razorpay-IN and axis-IN. That content will get skipped if we don't check for it. 
Logic for this is in docs-website `webpack-loaders/remark-plugins/remark-plugin-whitelabel/index.js`. Might need to duplicate that code here.

[the remark this answers is not in the document]
Comment (PranjalMaithani) [reply scripts/at-rules.js:85]: > For now, this attribute is only used for razorpay-IN org and axis org. 

Why is this not being used for Curlec?

**Review by PranjalMaithani — APPROVED** [review]
Approved with 2 comments
